consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/validation/RoundChangeCertificateValidator.java

```diff
@@ -0,0 +1,129 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft.validation;
+
+import static tech.pegasys.pantheon.consensus.ibft.IbftHelpers.findLatestPreparedCertificate;
+import static tech.pegasys.pantheon.consensus.ibft.IbftHelpers.prepareMessageCountForQuorum;
+
+import tech.pegasys.pantheon.consensus.ibft.ConsensusRoundIdentifier;
+import tech.pegasys.pantheon.consensus.ibft.IbftBlockHashing;
+import tech.pegasys.pantheon.consensus.ibft.IbftBlockInterface;
+import tech.pegasys.pantheon.consensus.ibft.IbftHelpers;
+import tech.pegasys.pantheon.consensus.ibft.payload.PreparedCertificate;
+import tech.pegasys.pantheon.consensus.ibft.payload.RoundChangeCertificate;
+import tech.pegasys.pantheon.consensus.ibft.payload.RoundChangePayload;
+import tech.pegasys.pantheon.consensus.ibft.payload.SignedData;
+import tech.pegasys.pantheon.consensus.ibft.validation.RoundChangePayloadValidator.MessageValidatorForHeightFactory;
+import tech.pegasys.pantheon.ethereum.core.Address;
+import tech.pegasys.pantheon.ethereum.core.Block;
+
+import java.util.Collection;
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class RoundChangeCertificateValidator {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Collection<Address> validators;
+  private final MessageValidatorForHeightFactory messageValidatorFactory;
+  private final long quorum;
+  private final long chainHeight;
+
+  public RoundChangeCertificateValidator(
+      final Collection<Address> validators,
+      final MessageValidatorForHeightFactory messageValidatorFactory,
+      final long chainHeight) {
+    this.validators = validators;
+    this.messageValidatorFactory = messageValidatorFactory;
+    this.quorum = IbftHelpers.calculateRequiredValidatorQuorum(validators.size());
+    this.chainHeight = chainHeight;
+  }
+
+  public boolean validateRoundChangeMessagesAndEnsureTargetRoundMatchesRoot(
+      final ConsensusRoundIdentifier expectedRound, final RoundChangeCertificate roundChangeCert) {
+
+    final Collection<SignedData<RoundChangePayload>> roundChangeMsgs =
+        roundChangeCert.getRoundChangePayloads();
+
+    if (roundChangeMsgs.size() < quorum) {
+      LOG.info("Invalid RoundChangeCertificate, insufficient RoundChange messages.");
+      return false;
+    }
+
+    if (!roundChangeCert.getRoundChangePayloads().stream()
+        .allMatch(p -> p.getPayload().getRoundIdentifier().equals(expectedRound))) {
+      LOG.info(
+          "Invalid RoundChangeCertificate, not all embedded RoundChange messages have a "
+              + "matching target round.");
+      return false;
+    }
+
+    for (final SignedData<RoundChangePayload> roundChangeMsg :
```

Comment:
The RoundChangePayloadValidator only needs to be created once. And this validation check could be replaced with allMatch on the stream so could do something like if (!roundChangeCert.getRoundChangePayloads().stream()
        .allMatch(roundChangeValidator::validateRoundChange)) { .. }

Reply:
done.